Simplenote/NSString+Simplenote.swift

```diff
@@ -5,6 +5,20 @@ import Foundation
 //
 extension NSString {
 
+    /// Percent Encodes all of the non alphanumeric characters in the receiver
+    ///
+    @objc
+    var byEncodingNonAlphanumerics: String? {
+        addingPercentEncoding(withAllowedCharacters: .alphanumerics)
+    }
+
+    /// Returns a Percent Encoded / lowercased and UTF normalized version of the receiver
+    ///
+    @objc
+    var byEncodingForTagSimperiumKey: String {
+        precomposedStringWithCanonicalMapping.lowercased().byEncodingNonAlphanumerics ?? self as String
```

Comment:
this is cool, though would it not be helpful to share some terms with the other apps concerning this transformation?

we could call this `asTagHash` or just `tagHash` or `tagHashOf`

I also find that separating `byEncodingNonAlphanumerics` and its corresponding test a bit fluffy, exposed externally when they are purely internal details to the tag hash algorithm. what's the purpose of separating them?

in any case, we don't really care if `byEncodingNonAlphanumerics` does that - we care if our tag names collapse to the appropriate cross-platform tag hash and so I would have expected any test of this behavior to test that instead of an internal implementation detail

Reply:
Mmmmmmmmmmm I don't like the name `TagHash` honestly. I can't tell why, though!! 😂 

Now, strictly speaking, this API does behave like a hashing method (unless I'm missing something big, mapping should always be 1:1). And sharing terminology could be useful eventually.

I'll get this one updated, thank you for the suggestion!

Reply:
> I don't like the name TagHash honestly. I can't tell why, though!! 😂

If you come up with a better one we can all harmonize on it. I was looking for something descriptive and terse.

> this API does behave like a hashing method

That's exactly what it's doing.

> unless I'm missing something big, mapping should always be 1:1

Almost every hashing routine will have collisions by design since hashing is a kind of collapsing of a dimension of data. If this weren't the case our MD5 and SHA hashes would have to have an infinite number of bits since they hash every possible sequence of bytes into 16 or 20 or so bytes.

a zero-collision mapping is called _perfect_ hashing while a 1:1 mapping is a _minimal perfect_ hash. these are rare but used when the input domain is finite and static and we really want to have practically instant lookup for the keys or when we need to save as much storage space as possible.

Reply:
And now I'm terribly curious as per whether we can find a way to make this API collision!